File: preprocess/source_locator.py
"""Source locator -- find containing elements and compute Lagrange weights.

Given a source position (x_s, y_s, z_s), locate all elements that contain
the source via Newton iteration in natural coordinates, then compute
Lagrange interpolation weights.

Two modes:
- **Surface mode** (default): source_z == z_min, searches only free-surface
  elements (boundary_tag == 1). The source may lie on a shared face, edge,
  or vertex -- all containing elements are returned with normalized weights.
- **Buried mode** (is_pml provided, source_z != z_min): searches all non-PML
  elements via AABB bounding box test + Newton iteration. Source should be
  inside a single element; >1 elements triggers a warning.

These precomputed weights and element list are stored in config.h5 so
the forward solver can inject the source without runtime search.
"""

from __future__ import annotations

import logging
import numpy as np
import numpy.typing as npt

from preprocess.gll_geometry import gll_quadrature_points
from preprocess.topology_reader import TopologyData

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Newton iteration for point-in-hexahedron
# ---------------------------------------------------------------------------

# Linear shape functions for 8-node hex (same as gll_geometry._linear_shape_derivs)
HEX_REF_CORNERS_LOC = np.array(
    [
        [-1, -1, -1],
        [1, -1, -1],
        [1, 1, -1],
        [-1, 1, -1],
        [-1, -1, 1],
        [1, -1, 1],
        [1, 1, 1],
        [-1, 1, 1],
    ],
    dtype=np.float64,
)


def _linear_shape_derivs_loc(
    xi: float, eta: float, zeta: float
) -> tuple[npt.NDArray[np.float64], npt.NDArray[np.float64]]:
    """Linear hex shape functions and their derivs at (xi, eta, zeta).

    Returns (N_vals [8], dN [8,3]) where dN[a,m] = dN_a/dxi_m.
    """
    N_vals = np.zeros(8, dtype=np.float64)
    dN = np.zeros((8, 3), dtype=np.float64)
    for a in range(8):
        ca, cb, cc = HEX_REF_CORNERS_LOC[a]
        t0 = 0.125
        N_vals[a] = t0 * (1 + ca * xi) * (1 + cb * eta) * (1 + cc * zeta)
        dN[a, 0] = t0 * ca * (1 + cb * eta) * (1 + cc * zeta)
        dN[a, 1] = t0 * (1 + ca * xi) * cb * (1 + cc * zeta)
        dN[a, 2] = t0 * (1 + ca * xi) * (1 + cb * eta) * cc
    return N_vals, dN


def _newton_find_xi(
    x_target: npt.NDArray[np.float64],
    corners: npt.NDArray[np.float64],
    max_iter: int = 50,
    tol: float = 1e-12,
) -> npt.NDArray[np.float64] | None:
    """Newton iteration to find natural coords (xi,eta,zeta) of a physical point.

    Uses the linear hex mapping: x(xi,eta,zeta) = sum N_a(xi,eta,zeta) * x_a.

    Args:
        x_target: [3] physical point.
        corners:  [8, 3] physical corner coordinates in GMSH order.
        max_iter: Maximum Newton iterations.
        tol:      Convergence tolerance on |Deltaxi|.

    Returns:
        [3] (xi, eta, zeta) in reference domain, or None if not converged.
    """
    xi = np.array([0.0, 0.0, 0.0], dtype=np.float64)  # start at element center

    for _ in range(max_iter):
        _N, dN = _linear_shape_derivs_loc(xi[0], xi[1], xi[2])
        x_current = _N @ corners  # [3] current physical estimate
        J = dN.T @ corners  # [3, 3] dx/dxi

        residual = x_target - x_current
        try:
            dxi = np.linalg.solve(J, residual)
        except np.linalg.LinAlgError:
            return None

        xi = xi + dxi
        if np.linalg.norm(dxi) < tol:
            break
    else:
        # Did not converge within max_iter
        return None

    # Check that the result is inside (or very close to) [-1, 1]^3
    margin = 1e-8
    if np.any(xi < -(1.0 + margin)) or np.any(xi > 1.0 + margin):
        return None

    return xi


# ---------------------------------------------------------------------------
# Lagrange interpolation weights at natural coordinates
# ---------------------------------------------------------------------------


def _lagrange_basis_at(xi_eval: float, nodes: npt.NDArray[np.float64]) -> npt.NDArray[np.float64]:
    """Evaluate 1-D Lagrange basis on given nodal points at xi_eval.

    Args:
        xi_eval: Evaluation point in [-1, 1].
        nodes:   Nodal positions [NGLL] (e.g., GLL quadrature points).

    Returns:
        w_i: [NGLL] Lagrange basis values l_i(xi_eval).
    """
    n = len(nodes)
    w = np.ones(n, dtype=np.float64)
    for i in range(n):
        for j in range(n):
            if i != j:
                w[i] *= (xi_eval - nodes[j]) / (nodes[i] - nodes[j])
    return w


def compute_source_weights(
    xi_vec: npt.NDArray[np.float64], gll_pts: npt.NDArray[np.float64]
) -> npt.NDArray[np.float64]:
    """Compute 3-D Lagrange interpolation weights at natural coords.

    w_ijk = l_i(xi_s) * l_j(eta_s) * l_k(zeta_s)

    Args:
        xi_vec: [3] natural coords (xi, eta, zeta).
        gll_pts: [NGLL] 1-D GLL nodal points.

    Returns:
        weights: [NGLL, NGLL, NGLL] Lagrange interpolation weights.
    """
    NGLL = len(gll_pts)
    lx = _lagrange_basis_at(xi_vec[0], gll_pts)
    ly = _lagrange_basis_at(xi_vec[1], gll_pts)
    lz = _lagrange_basis_at(xi_vec[2], gll_pts)
    return (
        lx[:, np.newaxis, np.newaxis]
        * ly[np.newaxis, :, np.newaxis]
        * lz[np.newaxis, np.newaxis, :]
    )


# ---------------------------------------------------------------------------
# Bounding box helper for buried source candidate search
# ---------------------------------------------------------------------------


def _find_candidate_elements(
    source_xyz: npt.NDArray[np.float64],
    gll_coords: npt.NDArray[np.float64],
    is_pml: npt.NDArray[np.bool_] | None = None,
) -> list[int]:
    """Find elements whose AABB contains source_xyz.

    Uses the GLL node bounding box (min/max over all GLL nodes) as a fast
    rejection filter before Newton iteration.

    Args:
        source_xyz:  [3] source position.
        gll_coords:  [n_cell, NGLL, NGLL, NGLL, 3] GLL node positions.
        is_pml:      [n_cell] bool mask, True for PML elements to exclude.

    Returns:
        List of element indices (0-based) whose AABB contains source_xyz.
    """
    n_cell = gll_coords.shape[0]
    candidates: list[int] = []

    # Compute per-element AABB once
    # gll_coords shape: [n_cell, NGLL, NGLL, NGLL, 3]
    x_min = gll_coords[..., 0].min(axis=(1, 2, 3))
    x_max = gll_coords[..., 0].max(axis=(1, 2, 3))
    y_min = gll_coords[..., 1].min(axis=(1, 2, 3))
    y_max = gll_coords[..., 1].max(axis=(1, 2, 3))
    z_min = gll_coords[..., 2].min(axis=(1, 2, 3))
    z_max = gll_coords[..., 2].max(axis=(1, 2, 3))

    margin = 1e-10
    sx, sy, sz = float(source_xyz[0]), float(source_xyz[1]), float(source_xyz[2])

    for e in range(n_cell):
        if is_pml is not None and bool(is_pml[e]):
            continue
        if (
            x_min[e] - margin <= sx <= x_max[e] + margin
            and y_min[e] - margin <= sy <= y_max[e] + margin
            and z_min[e] - margin <= sz <= z_max[e] + margin
        ):
            candidates.append(e)

    return candidates


# ---------------------------------------------------------------------------
# Corner extraction for Newton iteration (GMSH hex ordering)
# ---------------------------------------------------------------------------


def _get_element_corners(
    gll_coords: npt.NDArray[np.float64], e: int, ngll: int
) -> npt.NDArray[np.float64]:
    """Extract the 8 corner vertices of element e from GLL coords."""
    idx = ngll - 1
    corners = [
        gll_coords[e, 0, 0, 0],
        gll_coords[e, idx, 0, 0],
        gll_coords[e, idx, idx, 0],
        gll_coords[e, 0, idx, 0],
        gll_coords[e, 0, 0, idx],
        gll_coords[e, idx, 0, idx],
        gll_coords[e, idx, idx, idx],
        gll_coords[e, 0, idx, idx],
    ]
    return np.array(corners, dtype=np.float64)


# ---------------------------------------------------------------------------
# Main entry point: locate source (surface or buried)
# ---------------------------------------------------------------------------


def locate_source(
    topology: TopologyData,
    source_xyz: npt.NDArray[np.float64],
    gll_coords: npt.NDArray[np.float64],
    boundary_tag: npt.NDArray[np.int64],
    N: int,
    is_pml: npt.NDArray[np.bool_] | None = None,
) -> dict:
    """Locate source position, return element list + Lagrange weights.

    Two modes determined automatically:
    - **Surface mode**: when is_pml is None (legacy call) or source_z is
      the domain's z_min. Searches only free-surface elements (boundary_tag==1).
    - **Buried mode**: when is_pml is provided and source_z != z_min. Searches
      all non-PML elements via AABB bounding box + Newton iteration.

    Args:
        topology:    Mesh topology data.
        source_xyz:  [3] source position (x, y, z).
        gll_coords:  [n_cell, NGLL, NGLL, NGLL, 3] GLL node positions.
        boundary_tag: [n_surface] surface tags (0=interior, 1=free, 2=absorbing).
        N:           Polynomial order.
        is_pml:      [n_cell] bool mask, True for PML elements. Required for
                     buried mode; ignored in surface mode.

    Returns:
        dict with:
          element_ids: [n_src_elem] 0-based element indices (match partition local_element_ids).
          xi:          [n_src_elem] xi natural coordinate.
          eta:         [n_src_elem] eta natural coordinate.
          zeta:        [n_src_elem] zeta natural coordinate.
          weights:     [n_src_elem, NGLL, NGLL, NGLL] Lagrange w_ijk (normalized).
          n_src_elem:  int count.
          mode:        str, "surface" or "buried".
    """
    n_cell = topology.n_cell
    c2s = topology.cell_to_surface
    NGLL = N + 1
    gll_pts = gll_quadrature_points(N)

    # Determine mode
    # Surface mode: is_pml is None (legacy) OR source_z matches z_min
    zmin = float(gll_coords[:, :, :, :, 2].min())
    is_buried = is_pml is not None and abs(float(source_xyz[2]) - zmin) > 1e-8 * abs(zmin or 1.0)

    if is_buried:
        # ── Buried mode: search all non-PML elements ──
        if is_pml is None:
            raise ValueError("Buried mode requires is_pml array.")
        logger.debug(
            "Buried source at z=%.4f m, searching non-PML elements...", float(source_xyz[2])
        )
        candidates = _find_candidate_elements(source_xyz, gll_coords, is_pml)
        if not candidates:
            raise ValueError(
                f"Buried source at {source_xyz} not contained in any non-PML element."
            )
        element_iter = candidates
    else:
        # ── Surface mode: search free-surface elements ──
        logger.debug("Surface source at z=%.4f m, searching free-surface elements...", zmin)
        free_surface_cells: list[int] = []
        for e in range(n_cell):
            for fi in range(6):
                sid = abs(int(c2s[e, fi])) - 1
                if boundary_tag[sid] == 1:
                    free_surface_cells.append(e)
                    break

        if not free_surface_cells:
            raise ValueError(
                "No free-surface elements found. "
                "Check that the mesh has at least one surface with z asymp z_min."
            )
        element_iter = free_surface_cells

    # Common: Newton iteration and weight computation for each candidate
    element_ids: list[int] = []
    xi_list: list[float] = []
    eta_list: list[float] = []
    zeta_list: list[float] = []
    weights_list: list[npt.NDArray[np.float64]] = []

    for e in element_iter:
        corners = _get_element_corners(gll_coords, e, NGLL)

        xi_vec = _newton_find_xi(source_xyz, corners)
        if xi_vec is None:
            continue

        w = compute_source_weights(xi_vec, gll_pts)

        # Check that weights sum is reasonable (should be ~1 for a single elem)
        w_sum = np.sum(w)
        if w_sum < 1e-12:
            continue

        element_ids.append(e)  # 0-based element index (matches partition local_element_ids)
        xi_list.append(float(xi_vec[0]))
        eta_list.append(float(xi_vec[1]))
        zeta_list.append(float(xi_vec[2]))
        weights_list.append(w)

    if not element_ids:
        raise ValueError(
            f"Source at {source_xyz} not contained in any candidate element. "
            f"Checked {len(element_iter)} candidate(s)."
        )

    if is_buried and len(element_ids) > 1:
        logger.warning(
            "Buried source at %s found in %d element(s) — may lie on element "
            "face/edge. Verify source position is correct.",
            source_xyz,
            len(element_ids),
        )

    # Normalize weights across all containing elements
    # (mathematically required: scatter_to_rank is additive, so shared nodes
    #  would get Nx amplitude without normalization)
    total_weight = float(sum(np.sum(w) for w in weights_list))
    weights_norm = [w / total_weight for w in weights_list]

    return {
        "element_ids": np.array(element_ids, dtype=np.int64),
        "xi": np.array(xi_list, dtype=np.float64),
        "eta": np.array(eta_list, dtype=np.float64),
        "zeta": np.array(zeta_list, dtype=np.float64),
        "weights": np.array(weights_norm, dtype=np.float64),
        "n_src_elem": len(element_ids),
        "mode": "buried" if is_buried else "surface",
    }

File: preprocess/cpp/source_locator.cpp
/// source_locator.cpp — find source element(s) and compute Lagrange weights
///
/// Port of preprocess/source_locator.py.
/// Uses Eigen for small-matrix linear algebra (already a project dependency).

#include <Eigen/Dense>
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <vector>

#include "gf_config.h"

namespace gf {
namespace {

using Vec3 = Eigen::Vector3d;
using Mat3 = Eigen::Matrix3d;
using Arr8x3 = Eigen::Matrix<double, 8, 3>;

// ── GLL quadrature points (1-D) ────────────────────────────────────────────

/// Compute Gauss-Lobatto-Legendre nodes in [-1, 1].
/// Uses Newton iteration on the GLL derivative polynomial.
std::vector<double> gll_points(int N) {
    // Gauss-Lobatto-Legendre nodes in [-1, 1].
    // Uses Newton iteration to find the zeros of P_N'(x) (interior nodes).
    int n = N + 1;
    std::vector<double> xi(n);
    if (n == 1) {
        xi[0] = 0.0;
        return xi;
    }
    if (n == 2) {
        xi[0] = -1.0;
        xi[1] = 1.0;
        return xi;
    }
    xi[0] = -1.0;
    xi[n - 1] = 1.0;

    // Chebyshev initial guess for interior points
    for (int i = 1; i < n - 1; ++i) {
        xi[i] = -std::cos(M_PI * i / N);
    }

    // Newton iteration: find zeros of P_N'(x)
    // P_N'(x) = N * (x*P_N(x) - P_{N-1}(x)) / (x^2 - 1)
    // P_N''(x) = (2*x*P_N'(x) - N*(N+1)*P_N(x)) / (1 - x^2)
    for (int iter = 0; iter < 30; ++iter) {
        double max_delta = 0.0;
        for (int i = 1; i < n - 1; ++i) {
            double x = xi[i];

            // Evaluate P_{N-1}(x) and P_N(x) via recurrence
            double pn = 1.0;  // P_0
            double pn1 = x;   // P_1
            for (int k = 2; k <= N; ++k) {
                double pk = ((2.0 * k - 1.0) * x * pn1 - (k - 1.0) * pn) / k;
                pn = pn1;
                pn1 = pk;
            }
            // pn = P_{N-1}(x), pn1 = P_N(x)

            double x2 = x * x;
            double pprime = N * (x * pn1 - pn) / (x2 - 1.0);
            double p2prime = (2.0 * x * pprime - N * (N + 1.0) * pn1) / (1.0 - x2);
            double delta = pprime / p2prime;

            xi[i] -= delta;
            max_delta = std::max(max_delta, std::abs(delta));
        }
        if (max_delta < 1e-15)
            break;
    }
    return xi;
}

// ── Hex reference corners (GMSH ordering) ──────────────────────────────────

const Arr8x3& hex_corners() {
    static const Arr8x3 c = (Arr8x3() << -1, -1, -1, 1, -1, -1, 1, 1, -1, -1, 1, -1, -1, -1, 1, 1,
                             -1, 1, 1, 1, 1, -1, 1, 1)
                                .finished();
    return c;
}

// ── Newton iteration: find (xi,eta,zeta) for a physical point ─────────────

Vec3 newton_find_xi(const Vec3& target, const Arr8x3& corners, int max_iter = 50,
                    double tol = 1e-12) {
    Vec3 xi(0.0, 0.0, 0.0);  // start at element center

    for (int iter = 0; iter < max_iter; ++iter) {
        // Linear shape functions and derivatives at xi
        Eigen::Matrix<double, 8, 1> N;
        Eigen::Matrix<double, 8, 3> dN;
        const auto& c = hex_corners();
        for (int a = 0; a < 8; ++a) {
            double ca = c(a, 0), cb = c(a, 1), cc = c(a, 2);
            N(a) = 0.125 * (1 + ca * xi[0]) * (1 + cb * xi[1]) * (1 + cc * xi[2]);
            dN(a, 0) = 0.125 * ca * (1 + cb * xi[1]) * (1 + cc * xi[2]);
            dN(a, 1) = 0.125 * (1 + ca * xi[0]) * cb * (1 + cc * xi[2]);
            dN(a, 2) = 0.125 * (1 + ca * xi[0]) * (1 + cb * xi[1]) * cc;
        }

        Vec3 x_current = corners.transpose() * N;
        Mat3 J = corners.transpose() * dN;  // dx/dxi

        Vec3 residual = target - x_current;
        Vec3 dxi = J.colPivHouseholderQr().solve(residual);
        xi += dxi;
        if (dxi.norm() < tol)
            break;
    }

    return xi;
}

/// Check if xi is inside [-1,1]^3 within margin
bool inside_element(const Vec3& xi, double margin = 1e-8) {
    return xi[0] >= -(1.0 + margin) && xi[0] <= 1.0 + margin && xi[1] >= -(1.0 + margin) &&
           xi[1] <= 1.0 + margin && xi[2] >= -(1.0 + margin) && xi[2] <= 1.0 + margin;
}

// ── Lagrange basis at a point ──────────────────────────────────────────────

void lagrange_basis_at(double xi_eval, const std::vector<double>& nodes, std::vector<double>& w) {
    int n = static_cast<int>(nodes.size());
    w.assign(n, 1.0);
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            if (i != j) {
                w[i] *= (xi_eval - nodes[j]) / (nodes[i] - nodes[j]);
            }
        }
    }
}

/// Compute 3-D Lagrange interpolation weights at (xi, eta, zeta)
void compute_source_weights_3d(const Vec3& xi_vec, const std::vector<double>& gll_pts,
                               std::vector<double>& weights, int& ngll) {
    ngll = static_cast<int>(gll_pts.size());
    std::vector<double> lx, ly, lz;
    lagrange_basis_at(xi_vec[0], gll_pts, lx);
    lagrange_basis_at(xi_vec[1], gll_pts, ly);
    lagrange_basis_at(xi_vec[2], gll_pts, lz);

    weights.resize(ngll * ngll * ngll);
    for (int i = 0; i < ngll; ++i)
        for (int j = 0; j < ngll; ++j)
            for (int k = 0; k < ngll; ++k)
                weights[(i * ngll + j) * ngll + k] = lx[i] * ly[j] * lz[k];
}

}  // anonymous namespace

// ── Public API ──────────────────────────────────────────────────────────────

SourceResult locate_source(const Config& cfg, const double* gll_coords_flat, int n_cell, int ngll,
                           const int64_t* cell_to_surface, int n_surface,
                           const int64_t* boundary_tag, const int* is_pml) {
    SourceResult result;

    // User-specified source coordinates in meters.
    Vec3 source_pt(cfg.source_x_m, cfg.source_y_m, cfg.source_z_m >= 0 ? cfg.source_z_m : 0.0);
    bool is_buried = cfg.source_z_m >= 0;

    // Determine candidate elements
    std::vector<int> candidates;
    if (is_buried && is_pml) {
        // Buried mode: AABB test on non-PML elements
        for (int e = 0; e < n_cell; ++e) {
            if (is_pml[e])
                continue;
            // Compute AABB from GLL coords
            double xmin = 1e30, xmax = -1e30, ymin = 1e30, ymax = -1e30, zmin = 1e30, zmax = -1e30;
            int stride = ngll * ngll * ngll * 3;
            const double* ec = gll_coords_flat + e * stride;
            for (int i = 0; i < ngll * ngll * ngll; ++i) {
                xmin = std::min(xmin, ec[i * 3 + 0]);
                xmax = std::max(xmax, ec[i * 3 + 0]);
                ymin = std::min(ymin, ec[i * 3 + 1]);
                ymax = std::max(ymax, ec[i * 3 + 1]);
                zmin = std::min(zmin, ec[i * 3 + 2]);
                zmax = std::max(zmax, ec[i * 3 + 2]);
            }
            double margin = 1e-10;
            if (source_pt[0] >= xmin - margin && source_pt[0] <= xmax + margin &&
                source_pt[1] >= ymin - margin && source_pt[1] <= ymax + margin &&
                source_pt[2] >= zmin - margin && source_pt[2] <= zmax + margin)
                candidates.push_back(e);
        }
    } else {
        // Surface mode: search free-surface elements
        for (int e = 0; e < n_cell; ++e) {
            for (int fi = 0; fi < 6; ++fi) {
                int sid = std::abs(static_cast<int>(cell_to_surface[e * 6 + fi])) - 1;
                if (sid >= 0 && sid < n_surface && boundary_tag[sid] == 1) {
                    candidates.push_back(e);
                    break;
                }
            }
        }
    }
    if (candidates.empty()) {
        fprintf(stderr, "ERROR: source not found in any candidate element\n");
        std::exit(1);
    }

    // GLL points
    auto gll_pts = gll_points(ngll - 1);

    // Newton iteration for each candidate
    int stride = ngll * ngll * ngll * 3;
    for (int e : candidates) {
        // Extract corners from GLL coords
        Arr8x3 corners;
        int idx = ngll - 1;
        const double* ec = gll_coords_flat + e * stride;
        auto get_node = [&](int i, int j, int k) -> Vec3 {
            int off = (i * ngll * ngll + j * ngll + k) * 3;
            return Vec3(ec[off], ec[off + 1], ec[off + 2]);
        };
        corners.row(0) = get_node(0, 0, 0);
        corners.row(1) = get_node(idx, 0, 0);
        corners.row(2) = get_node(idx, idx, 0);
        corners.row(3) = get_node(0, idx, 0);
        corners.row(4) = get_node(0, 0, idx);
        corners.row(5) = get_node(idx, 0, idx);
        corners.row(6) = get_node(idx, idx, idx);
        corners.row(7) = get_node(0, idx, idx);

        Vec3 xi = newton_find_xi(source_pt, corners);
        if (!inside_element(xi))
            continue;

        int w_ngll;
        std::vector<double> w;
        compute_source_weights_3d(xi, gll_pts, w, w_ngll);

        double w_sum = 0.0;
        for (double v : w)
            w_sum += v;
        if (w_sum < 1e-12)
            continue;

        // Normalize
        for (double& v : w)
            v /= w_sum;

        result.cell_ids.push_back(e);
        result.xi.push_back(xi[0]);
        result.eta.push_back(xi[1]);
        result.zeta.push_back(xi[2]);
        result.weights.push_back(w);
    }

    if (result.cell_ids.empty()) {
        fprintf(stderr, "ERROR: source at (%g,%g,%g) not contained in any element\n", source_pt[0],
                source_pt[1], source_pt[2]);
        std::exit(1);
    }

    result.n_src_cell = static_cast<int>(result.cell_ids.size());
    return result;
}

}  // namespace gf